alerting/src/main/kotlin/org/opensearch/alerting/aggregation/bucketselectorext/BucketSelectorExtAggregator.kt

```diff
@@ -0,0 +1,168 @@
+    private var name: String? = null
+    private var bucketsPathsMap: Map<String, String>
+    private var parentBucketPath: String
+    private var script: Script
+    private var gapPolicy: GapPolicy
+    private var bucketSelectorExtFilter: BucketSelectorExtFilter? = null
+
+    constructor(
+        name: String?,
+        bucketsPathsMap: Map<String, String>,
+        parentBucketPath: String,
+        script: Script,
+        gapPolicy: GapPolicy,
+        filter: BucketSelectorExtFilter?,
+        metadata: Map<String, Any>?
+    ) : super(name, bucketsPathsMap.values.toTypedArray(), metadata) {
+        this.bucketsPathsMap = bucketsPathsMap
+        this.parentBucketPath = parentBucketPath
+        this.script = script
+        this.gapPolicy = gapPolicy
+        this.bucketSelectorExtFilter = filter
+    }
+
+    /**
+     * Read from a stream.
+     */
+    @Suppress("UNCHECKED_CAST")
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : super(sin.readString(), null, null) {
+        script = Script(sin)
+        gapPolicy = GapPolicy.readFrom(sin)
+        bucketsPathsMap = sin.readMap() as Map<String, String>
+        parentBucketPath = sin.readString()
+        if (sin.readBoolean()) {
+            bucketSelectorExtFilter = BucketSelectorExtFilter(sin)
+        } else {
+            bucketSelectorExtFilter = null
+        }
+    }
+
+    @Throws(IOException::class)
+    override fun doWriteTo(out: StreamOutput) {
+        out.writeString(name)
+        script.writeTo(out)
+        gapPolicy.writeTo(out)
+        out.writeGenericValue(bucketsPathsMap)
+        out.writeString(parentBucketPath)
+        if (bucketSelectorExtFilter != null) {
+            out.writeBoolean(true)
+            bucketSelectorExtFilter!!.writeTo(out)
+        } else {
+            out.writeBoolean(false)
+        }
+    }
+
+    override fun getWriteableName(): String {
+        return NAME.preferredName
+    }
+
+    override fun doReduce(aggregations: Aggregations, reduceContext: ReduceContext): InternalAggregation {
+        val parentBucketPathList = AggregationPath.parse(parentBucketPath).pathElementsAsStringList
+        var subAggregations: Aggregations = aggregations
+        for (i in 0 until parentBucketPathList.size - 1) {
+            subAggregations = subAggregations.get<SingleBucketAggregation>(parentBucketPathList[0]).aggregations
+        }
+        val originalAgg = subAggregations.get(parentBucketPathList.last()) as InternalMultiBucketAggregation<*, *>
+        val buckets = originalAgg.buckets
+        val factory = reduceContext.scriptService().compile(script, BucketAggregationSelectorScript.CONTEXT)
+        val selectedBucketsIndex: MutableList<Int> = ArrayList()
+        for (i in buckets.indices) {
+            val bucket = buckets[i]
+            if (bucketSelectorExtFilter != null) {
+                var accepted = true
+                if (bucketSelectorExtFilter!!.isCompositeAggregation) {
```

Comment:
Should we break this function into smaller pieces to make it more readable? Outsource the responsibility of updating the accepted variable to another function?

Reply:
Some of the inner loops could potentially be broken up. Will revisit this later though to give time to merge in other changes.